packages/core/src/actions/execute.ts

```diff
@@ -14,122 +13,274 @@ import {
+}
 
-  receipts.push(...executionReceipts)
+export type ExecuteTransaction = (
+  deploymentContext: DeploymentContext,
+  transaction: MinimumTransaction,
+  executionMode: ExecutionMode,
+  signer?: ethers.Signer
+) => Promise<TransactionReceipt>
 
-  if (status === MerkleRootStatus.FAILED) {
-    return { success: false, receipts, failureAction }
-  } else {
-    logger?.info(`[Sphinx]: executed actions`)
-  }
+export type InjectRoles = (
+  deploymentContext: DeploymentContext,
+  executionMode: ExecutionMode
+) => Promise<void>
 
-  // We're done!
-  logger?.info(`[Sphinx]: successfully deployed project`)
-  return { success: true, receipts }
+/**
+ * Before calling the `compileAndExecuteDeployment` function, we have to construct a DeploymentContext
+ * which implements a number of functions that vary depending on the specific context in which the function
+ * is executed. This allows us to share the `compileAndExecuteDeployment` function between the proposal
+ * simulation, deploy cli command simulation/execution, and the website execution flow.
+ *
+ * Using a single function for all of these related flows improves maintainability and also reduces the
+ * chance that a deployment will be proposed successfully and then later fail due to a bug that only exists
+ * in the execution logic on the website.
+ *
+ * So to acheive those benefits, we aim to share as much logic between these different execution cases by
+ * limiting the amount of logic that is implemented within the `DeploymentContext` object. We should *only*
+ * use the `DeploymentContext` object to store logic which absolutely cannot be used within all three flows.
+ *
+ */
+export type DeploymentContext = {
+  /**
+   * Used when we need to throw an error ourselves.
+   *
+   * We use the `DeploymentContext` for this because the website needs to log errors to sentry
+   * while the propose and deploy commands should simply throw an error.
+   */
+  throwError: ThrowError
+
+  /**
+   * Used when catching an error throw by some dependency or function call.
+   *
+   * We use the `DeploymentContext` for this because the website needs to log errors to sentry
+   * while the propose and deploy commands should simply throw an error.
+   */
+  handleError: HandleError
+
+  /**
+   * Handles if the deployment has already been executed.
+   *
+   * We use the `DeploymentContext` for this because the website implements retry logic, as a result
+   * there are some cases where the deployment may be partially or fully executed already when we call
+   * the `compileAndExecuteDeployment` function. For the propose/deploy commands, this should never happen
+   * so we just error.
+   *
+   * It's worth noting that if we wanted to implement retry logic in the deploy cli command to make it more
+   * reliable, we would want to do that by implementing this function and then repeatedly calling the
+   * `compileAndExecuteDeployment` function. This is how the website currently works.
```

Comment:
We'll probably still want to have this in the context object even after implementing retries in the simulation/deploy command because the website handles retries by writing info to the database. So unless we're going to use a local db like sqlite, it probably wouldn't make very much sense to share that logic.